[AutoDiff] Disable `differentiable_function_extract` explicit type as…

dan-zheng opened the pull request:
…sertion.

`differentiability_function_extract` instruction has an optional explicit
extractee type. This is currently used by TypeSubstCloner and the LoadableByAddress
transform to rewrite `differentiability_function_extract` instructions while
preserving `@differentiable` function type invariants.

There is an assertion that `differentiability_function_extract` instructions do
not have explicit extractee types outside of canonical/lowered SIL. However,
this does not handle the SIL deserialization case above: when a function
containing a `differentiable_function_extract` instruction with an explicit type
is deserialized into a raw SIL module (which happens when optimizations are
enabled).

Removing the assertion unblocks this encountered use case.

Resolves SR-14004.

dan-zheng committed:
[AutoDiff] Disable `differentiable_function_extract` explicit type assertion.

`differentiability_function_extract` instruction has an optional explicit
extractee type. This is currently used by TypeSubstCloner and the LoadableByAddress
transform to rewrite `differentiability_function_extract` instructions while
preserving `@differentiable` function type invariants.

There is an assertion that `differentiability_function_extract` instructions do
not have explicit extractee types outside of canonical/lowered SIL. However,
this does not handle the SIL deserialization case above: when a function
containing a `differentiable_function_extract` instruction with an explicit type
is deserialized into a raw SIL module (which happens when optimizations are
enabled).

Removing the assertion unblocks this encountered use case.

A more robust longer-term solution may be to change SIL `@differentiable`
function types to explicitly store component original/JVP/VJP function types.

Resolves SR-14004.

dan-zheng commented on lib/SIL/IR/SILInstructions.cpp:
```diff
@@ -727,18 +727,6 @@ DifferentiableFunctionExtractInst::DifferentiableFunctionExtractInst(
-            module.getStage() == SILStage::Lowered) &&
-           "Explicit type is valid only in canonical or lowered SIL");
-  }
-#endif
```
Relaxing this assertion isn't so nice, but I would appreciate being able to land it to unblock a use case unless a different solution is apparent.

A more robust longer-term solution may be to change SIL `@differentiable` function types to explicitly store component original/JVP/VJP function types, instead of inferring JVP/VJP types from the original type. @rxwei had this idea long ago.

That would seem to solve this issue, along with enabling the following:
- Robustly re-enabling `differentiable_function` instruction verification (TF-1197).
- Enabling straightforward cloning/transformation of `differentiable_function_extract`
  instructions without [worrying about breaking invariants](https://github.com/apple/swift/blob/949b0c0cc3632b0c278a33a4022dd280cf31215a/include/swift/SIL/TypeSubstCloner.h#L325-L348).

dan-zheng commented:
@swift-ci Please test

marcrasi reviewed (approved):
Looks reasonable for a short term fix to me.

Could you merge this into https://github.com/google/swift/tree/holiday-patches ? This is the branch that we are using to collect and build all our PRs while main is frozen.

dan-zheng committed:
[AutoDiff] Fix `differentiable_function_extract` extractee type serialization.

The explicit extractee type was previously not serialized.
The operand's type was incorrectly used instead.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test
